File: app/src/main/java/com/neko/tools/SpeedTestActivity.java
package com.neko.tools;

import android.graphics.Bitmap;
import android.os.Bundle;
import android.view.View;
import android.view.KeyEvent;
import android.webkit.WebView;
import android.webkit.WebViewClient;
import android.widget.ProgressBar;
import android.widget.TextView;
import com.v2ray.ang.ui.BaseActivity;
import com.v2ray.ang.R;

public class SpeedTestActivity extends BaseActivity {
    ProgressBar progressBar;
    TextView txtwait;
    WebView web;

    public class myWebClient extends WebViewClient {
        private final SpeedTestActivity this$0;

        public myWebClient(SpeedTestActivity speedTestActivity) {
            this.this$0 = speedTestActivity;
        }

        @Override
        public void onPageFinished(WebView webView, String str) {
            super.onPageFinished(webView, str);
            this.this$0.txtwait.setVisibility(View.GONE);
            this.this$0.progressBar.setVisibility(View.GONE);
        }

        @Override
        public void onPageStarted(WebView webView, String str, Bitmap bitmap) {
            super.onPageStarted(webView, str, bitmap);
        }

        @Override
        public boolean shouldOverrideUrlLoading(WebView webView, String str) {
            webView.loadUrl(str);
            return true;
        }
    }

    public void onCreate(Bundle bundle) {
        super.onCreate(bundle);
        setContentView(R.layout.uwu_speedtest);
        this.web = (WebView) findViewById(R.id.uwu_WebView);
        this.progressBar = (ProgressBar) findViewById(R.id.uwu_ProgressBar);
        this.txtwait = (TextView) findViewById(R.id.uwu_TextView);
        this.web.setWebViewClient(new myWebClient(this));
        this.web.getSettings().setJavaScriptEnabled(true);
        this.web.loadUrl("https://fusiontempest.speedtestcustom.com");
    }

    public boolean onKeyDown(int i, KeyEvent keyEvent) {
        if (i == 4 && this.web.canGoBack()) {
            this.web.goBack();
            return true;
        }
        return super.onKeyDown(i, keyEvent);
    }
}

File: V2rayNG/app/src/main/java/com/neko/tools/SpeedTestActivity.kt
package com.neko.tools

import android.graphics.Bitmap
import android.os.Bundle
import android.view.KeyEvent
import android.view.View
import android.webkit.WebSettings
import android.webkit.WebView
import android.webkit.WebViewClient
import android.widget.ProgressBar
import android.widget.TextView

import com.neko.v2ray.ui.BaseActivity
import com.neko.v2ray.R

class SpeedTestActivity : BaseActivity() {
    private lateinit var progressBar: ProgressBar
    private lateinit var txtwait: TextView
    private lateinit var web: WebView

    private inner class MyWebClient : WebViewClient() {
        override fun onPageFinished(webView: WebView, url: String) {
            super.onPageFinished(webView, url)
            txtwait.visibility = View.GONE
            progressBar.visibility = View.GONE
        }

        override fun onPageStarted(webView: WebView, url: String, favicon: Bitmap?) {
            super.onPageStarted(webView, url, favicon)
        }

        override fun shouldOverrideUrlLoading(webView: WebView, url: String): Boolean {
            webView.loadUrl(url)
            return true
        }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.uwu_speedtest)
        
        web = findViewById(R.id.uwu_WebView)
        progressBar = findViewById(R.id.uwu_ProgressBar)
        txtwait = findViewById(R.id.uwu_TextView)
        
        web.webViewClient = MyWebClient()
        web.settings.apply {
            javaScriptEnabled = true
            domStorageEnabled = true
            loadsImagesAutomatically = true
            mixedContentMode = WebSettings.MIXED_CONTENT_ALWAYS_ALLOW
            loadWithOverviewMode = true
            useWideViewPort = true
        }
        web.setInitialScale(1)
        web.clearCache(true)
        web.clearHistory()
        web.loadUrl("https://fusiontempest.speedtestcustom.com")
    }

    override fun onKeyDown(keyCode: Int, event: KeyEvent?): Boolean {
        if (keyCode == KeyEvent.KEYCODE_BACK && web.canGoBack()) {
            web.goBack()
            return true
        }
        return super.onKeyDown(keyCode, event)
    }
}
